bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/internal/scheduler2/InternalSchedulerImpl.java

```diff
@@ -0,0 +1,320 @@
+        });
+        return deferred;
+    }
+
+    static class Unique {
+        AtomicBoolean done = new AtomicBoolean();
+
+        interface RunnableException {
+            public void run();
+        }
+
+        boolean once(RunnableException o) {
+            if (done.getAndSet(true) == false) {
+                o.run();
+                return true;
+            } else {
+                return false;
+            }
+        }
+    }
+
+    @Override
+    public <T> CompletableFuture<T> before(CompletableFuture<T> promise, Duration timeout) {
+        CompletableFuture<T> d = new CompletableFuture<T>();
+        Unique only = new Unique();
+
+        after(timeout).thenAccept((p) -> {
+            only.once(() -> d.completeExceptionally(TimeoutException.SINGLETON));
+        });
+
+        promise.thenAccept((p) -> {
+            only.once(() -> d.complete(p));
+        }).exceptionally((Throwable ex) -> {
+            only.once(() -> d.completeExceptionally(ex));
+            return null;
+        });
+        return d;
+    }
+
+    private static abstract class Schedule {
+        volatile CompletableFuture<?> promise;
+        volatile boolean canceled;
+        Throwable exception;
+
+        abstract Instant next();
+
+        abstract void doIt() throws Exception;
+    }
+
+    private static class PeriodSchedule extends Schedule {
+        private long last;
+        private Iterator<Long> iterator;
+        private long rover;
+        private RunnableWithException runnable;
+
+        @Override
+        Instant next() {
+            if (iterator.hasNext()) {
+                last = iterator.next();
+            }
+
+            return Instant.ofEpochMilli(rover += last);
+        }
+
+        @Override
+        void doIt() throws Exception {
+            runnable.run();
+        }
+    }
+
+    @Override
+    public Closeable schedule(RunnableWithException r, Duration first, Duration... delays) {
+        PeriodSchedule s = new PeriodSchedule();
+        s.iterator = Arrays.stream(delays).map(delay -> delay.toMillis()).iterator();
+        s.runnable = r;
+        s.rover = System.currentTimeMillis() + first.toMillis();
+        s.last = first.toMillis();
+        schedule(s, Instant.ofEpochMilli(first.toMillis() + System.currentTimeMillis()));
+        return () -> {
+            s.canceled = true;
+            s.promise.cancel(true);
+        };
+    }
+
+    private void schedule(Schedule s, Instant instant) {
+        s.promise = at(() -> {
+            try {
+                s.doIt();
+            } catch (Throwable t) {
+                if (s.exception != null) {
+                    logger.warn("Schedule failed {}", s, t);
+                }
+                s.exception = t;
+            }
+
+            schedule(s, s.next());
+            return null;
+        }, instant);
+        if (s.canceled) {
+            s.promise.cancel(true);
+        }
+    }
+
+    private class ScheduleCron extends Schedule {
```

Comment:
I would generalise this class in order to use any TemporalAdjuster instead of a CronAdjuster